processor/src/debug.rs

```diff
@@ -313,3 +314,113 @@ impl AsRef<AssemblyOp> for AsmOpInfo {
         &self.asmop
     }
 }
+
+// BUS DEBUGGING
+// =================================================================
+
+/// A message that can be sent on a bus.
+pub(crate) trait BusMessage<E: FieldElement<BaseField = Felt>>: fmt::Display {
+    /// The concrete value that this message evaluates to.
+    fn value(&self, alphas: &[E]) -> E;
+
+    /// The source of this message (e.g. "mload" or "memory chiplet").
+    fn source(&self) -> &str;
+}
+
+/// Note: we use `Vec` internally instead of a `BTreeMap` as a workaround for field elements not
+/// implementing `Ord`. Since this is only used in debug/test code, this is acceptable.
+pub(crate) struct BusDebugger<E: FieldElement<BaseField = Felt>> {
+    pub bus_name: String,
+    pub outstanding_requests: Vec<(E, Box<dyn BusMessage<E>>)>,
+    pub outstanding_responses: Vec<(E, Box<dyn BusMessage<E>>)>,
+}
+
+impl<E> BusDebugger<E>
+where
+    E: FieldElement<BaseField = Felt>,
+{
+    pub fn new(bus_name: String) -> Self {
+        Self {
+            bus_name,
+            outstanding_requests: Vec::new(),
+            outstanding_responses: Vec::new(),
+        }
+    }
+}
+
+impl<E> BusDebugger<E>
+where
+    E: FieldElement<BaseField = Felt>,
+{
```

Comment:
This `impl` is needed only when the `testing` features is on, right? If so, we could just gate it with `testing` feature and then get rid of `#[allow(dead_code)]` attributes below.

Reply:
This results in a cascading effect:

- we then need to annotate the `Display` impl,
- which results in needing to annotate each field of the `BusDebugger`
- which results in needing to annotate the `new()` method,
- which results in a compiler error since `E` is no longer used in `BusDebugger` (when the `slow-testing` feature isn't on)

So I think it ends up being cleaner to just allow dead code